File: src/test/java/com/xpn/xwiki/doc/XWikiDocumentTest.java
/*
 * Copyright 2007, XpertNet SARL, and individual contributors as indicated
 * by the contributors.txt.
 *
 * This is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation; either version 2.1 of
 * the License, or (at your option) any later version.
 *
 * This software is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this software; if not, write to the Free
 * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
 * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
 */
package com.xpn.xwiki.doc;

import java.util.Date;

import org.jmock.cglib.MockObjectTestCase;
import org.jmock.Mock;
import com.xpn.xwiki.XWikiContext;
import com.xpn.xwiki.XWiki;
import com.xpn.xwiki.XWikiConfig;
import com.xpn.xwiki.XWikiException;
import com.xpn.xwiki.render.XWikiRenderingEngine;
import com.xpn.xwiki.store.XWikiHibernateVersioningStore;

/**
 * Unit tests for {@link XWikiDocument}.
 *
 * @version $Id: $
 */
public class XWikiDocumentTest extends MockObjectTestCase
{
    private XWikiContext context;
    private XWikiDocument document;
    private Mock mockXWiki;
    private Mock mockXWikiRenderingEngine;
    private Mock mockXWikiVersioningStore;
    
    protected void setUp()
    {
        this.context = new XWikiContext();
        this.document = new XWikiDocument("Space", "Page");

        this.mockXWiki = mock(XWiki.class, new Class[] {XWikiConfig.class, XWikiContext.class},
            new Object[] {new XWikiConfig(), this.context});
        this.mockXWiki.stubs().method("Param").will(returnValue(null));

        this.mockXWikiRenderingEngine = mock(XWikiRenderingEngine.class);
        
        this.mockXWikiVersioningStore =
            mock(XWikiHibernateVersioningStore.class, new Class[] {XWiki.class,
            XWikiContext.class}, new Object[] {this.mockXWiki.proxy(), this.context});
        this.mockXWikiVersioningStore.stubs().method("getXWikiDocumentArchive").will(
            returnValue(null));

        this.mockXWiki.stubs().method("getRenderingEngine").will(returnValue(
            this.mockXWikiRenderingEngine.proxy()));
        this.mockXWiki.stubs().method("getVersioningStore").will(
            returnValue(this.mockXWikiVersioningStore.proxy()));
        
        this.context.setWiki((XWiki) this.mockXWiki.proxy());
    }

    public void testGetDisplayTitleWhenNoTitleAndNoContent()
    {
        this.document.setContent("Some content");

        assertEquals("Page", this.document.getDisplayTitle(this.context));
    }

    public void testGetDisplayWhenTitleExists()
    {
        this.document.setContent("Some content");
        this.document.setTitle("Title");
        this.mockXWikiRenderingEngine.expects(once()).method("interpretText")
            .with(eq("Title"), ANYTHING, ANYTHING).will(returnValue("Title"));

        assertEquals("Title", this.document.getDisplayTitle(this.context));
    }

    public void testGetDisplayWhenNoTitleButSectionExists()
    {
        this.document.setContent("Some content\n1 Title");
        this.mockXWikiRenderingEngine.expects(once()).method("interpretText")
            .with(eq("Title"), ANYTHING, ANYTHING).will(returnValue("Title"));

        assertEquals("Title", this.document.getDisplayTitle(this.context));
    }
    
    public void testAuthorAfterDocumentCopy() throws XWikiException
    {
        String author = "Albatross";
        this.document.setAuthor(author);
        XWikiDocument copy =
            this.document.copyDocument(this.document.getName() + " Copy", this.context);

        assertTrue(author.equals(copy.getAuthor()));
    }

    public void testCreatorAfterDocumentCopy() throws XWikiException
    {
        String creator = "Condor";
        this.document.setCreator(creator);
        XWikiDocument copy =
            this.document.copyDocument(this.document.getName() + " Copy", this.context);

        assertTrue(creator.equals(copy.getCreator()));
    }
    
    public void testCreationDateAfterDocumentCopy() throws XWikiException, InterruptedException
    {
        Date sourceCreationDate = this.document.getCreationDate();
        Thread.sleep(1000);
        XWikiDocument copy =
            this.document.copyDocument(this.document.getName() + " Copy", this.context);
        
        assertTrue(copy.getCreationDate().equals(sourceCreationDate));
    }
}

File: xwiki-platform-core/xwiki-platform-node/src/main/node/core/document/document-xwiki/src/__tests__/XWikiDocument.test.ts
/**
 * See the NOTICE file distributed with this work for additional
 * information regarding copyright ownership.
 *
 * This is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation; either version 2.1 of
 * the License, or (at your option) any later version.
 *
 * This software is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this software; if not, write to the Free
 * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
 * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
 */

import { XWikiDocument } from "../XWikiDocument";
import { EntityType, Model } from "@xwiki/platform-xwiki-model-api";
import { afterEach, beforeEach, describe, expect, it, vi } from "vitest";
import type { EntityReference } from "@xwiki/platform-xwiki-model-api";
import type { XWikiMeta } from "@xwiki/platform-xwiki-utils";

function reference(representation: string): EntityReference {
  return Model.resolve(representation, EntityType.DOCUMENT)!;
}

function fakeMeta(overrides: Partial<XWikiMeta> = {}): XWikiMeta {
  return {
    documentReference: reference("xwiki:Space.Page"),
    version: "3.1",
    restURL: "/xwiki/rest/wikis/xwiki/spaces/Space/pages/Page",
    form_token: "token",
    isNew: false,
    locale: "",
    realLocale: "en",
    action: "edit",
    setVersion: vi.fn(),
    refreshVersion: vi.fn(),
    ...overrides,
  };
}

// A stand-in for the legacy API of the global XWiki object, recording the arguments it is called with.
const legacyAPI = {
  getURL: vi.fn(
    (action?: string, parameters?: URLSearchParams, fragment?: string) =>
      `/xwiki/bin/${action}/Space/Page?${parameters ?? ""}#${fragment ?? ""}`,
  ),
  getRestURL: vi.fn(
    (entity?: string, parameters?: URLSearchParams) =>
      `/xwiki/rest/wikis/xwiki/spaces/Space/pages/Page/${entity ?? ""}?${parameters ?? ""}`,
  ),
};

beforeEach(() => {
  vi.stubGlobal("XWiki", {
    Document: vi.fn(() => legacyAPI),
    currentDocument: legacyAPI,
    Model,
    EntityType,
  });
  document.body.innerHTML = `
    <input type="hidden" id="editingVersionDate" value="1700000000000"/>
    <input type="hidden" id="isNew" value="false"/>`;
});

afterEach(() => {
  vi.unstubAllGlobals();
  vi.clearAllMocks();
  document.body.innerHTML = "";
});

describe("currentDocument", () => {
  it("takes its fields from the meta information", () => {
    const meta = fakeMeta({ locale: "fr", isNew: true });
    const doc = XWikiDocument.currentDocument(meta);

    expect(doc.documentReference).toBe(meta.documentReference);
    expect(doc.language).toBe("fr");
    expect(doc.version).toBe("3.1");
    expect(doc.isNew).toBe(true);
    // The raw locale is not empty, so the real locale is not needed.
    expect(doc.translations).toBeUndefined();
  });

  it("takes the real locale from the meta information for the original translation", () => {
    const doc = XWikiDocument.currentDocument(fakeMeta({ realLocale: "de" }));

    expect(doc.language).toBe("");
    expect(doc.defaultLocale).toBe("de");
    expect(doc.realLocale).toBe("de");
  });

  it("instantiates the class it is called on", () => {
    class SubDocument extends XWikiDocument {}

    expect(SubDocument.currentDocument(fakeMeta())).toBeInstanceOf(SubDocument);
  });
});

describe("realLocale", () => {
  it("falls back on the default locale when the raw locale is empty", () => {
    const doc = new XWikiDocument({
      language: "",
      translations: { default: "en" },
    });

    expect(doc.realLocale).toBe("en");
  });

  it("is the root locale for a technical document", () => {
    const doc = new XWikiDocument({
      language: "",
      translations: { default: "" },
    });

    expect(doc.realLocale).toBe("");
  });

  it("is the raw locale for a translation", () => {
    const doc = new XWikiDocument({
      language: "fr",
      translations: { default: "en" },
    });

    expect(doc.realLocale).toBe("fr");
  });
});

describe("getRestURL", () => {
  it("targets the wiki page for the original translation", () => {
    new XWikiDocument({ language: "" }).getRestURL("objects");

    expect(legacyAPI.getRestURL).toHaveBeenCalledWith("objects", undefined);
  });

  it("targets the translation when the raw locale is set", () => {
    new XWikiDocument({ language: "fr" }).getRestURL("objects");

    expect(legacyAPI.getRestURL).toHaveBeenCalledWith(
      "translations/fr/objects",
      undefined,
    );
  });

  it("ignores the translation in getPageRestURL", () => {
    new XWikiDocument({ language: "fr" }).getPageRestURL("channels");

    expect(legacyAPI.getRestURL).toHaveBeenCalledWith("channels", undefined);
  });
});

describe("reload", () => {
  it("merges the REST response, ignoring its null properties", async () => {
    vi.stubGlobal(
      "fetch",
      vi.fn(async () =>
        Response.json({ version: "4.1", title: "Page", content: null }),
      ),
    );
    const doc = new XWikiDocument({
      version: "3.1",
      content: "before",
      isNew: true,
    });

    await doc.reload();

    expect(doc.version).toBe("4.1");
    expect(doc.title).toBe("Page");
    // The null content of the response must not overwrite the one we have.
    expect(doc.content).toBe("before");
    expect(doc.isNew).toBe(false);
  });

  it("resets the document when it doesn't exist anymore", async () => {
    vi.stubGlobal(
      "fetch",
      vi.fn(async () => new Response("", { status: 404 })),
    );
    const doc = new XWikiDocument({ version: "3.1", content: "before" });

    await doc.reload();

    expect(doc.version).toBe("1.1");
    expect(doc.modified).toBe(0);
    expect(doc.content).toBe("");
    expect(doc.isNew).toBe(true);
  });

  it("keeps the current data when the request fails", async () => {
    vi.stubGlobal(
      "fetch",
      vi.fn(async () => new Response("", { status: 500 })),
    );
    const doc = new XWikiDocument({ version: "3.1", content: "before" });

    await doc.reload();

    expect(doc.version).toBe("3.1");
    expect(doc.content).toBe("before");
  });
});

describe("syncCurrentDocumentState", () => {
  it("updates the meta version and the hidden fields of the edit form", () => {
    const meta = fakeMeta();
    const doc = XWikiDocument.currentDocument(meta);

    doc.update({ version: "4.1", modified: 1700000009999, isNew: false });

    expect(meta.setVersion).toHaveBeenCalledWith("4.1");
    expect(
      (document.getElementById("editingVersionDate") as HTMLInputElement).value,
    ).toBe("1700000009999");
    expect((document.getElementById("isNew") as HTMLInputElement).value).toBe(
      "false",
    );
  });

  it("does nothing when the version didn't change", () => {
    const meta = fakeMeta();
    const doc = XWikiDocument.currentDocument(meta);

    doc.update({ modified: 1700000009999 });

    expect(meta.setVersion).not.toHaveBeenCalled();
    expect(
      (document.getElementById("editingVersionDate") as HTMLInputElement).value,
    ).toBe("1700000000000");
  });

  it("leaves the page state alone for a document that is not the current one", () => {
    const meta = fakeMeta();
    const doc = XWikiDocument.currentDocument(meta);
    doc.documentReference = reference("xwiki:Space.Other");

    doc.update({ version: "4.1", modified: 1700000009999 });

    expect(doc.isCurrentDocument()).toBe(false);
    expect(meta.setVersion).not.toHaveBeenCalled();
    expect(
      (document.getElementById("editingVersionDate") as HTMLInputElement).value,
    ).toBe("1700000000000");
  });

  it("leaves the page state alone for a document built without the meta information", () => {
    const doc = new XWikiDocument({
      documentReference: reference("xwiki:Space.Page"),
      version: "3.1",
    });

    doc.update({ version: "4.1", modified: 1700000009999 });

    expect(doc.isCurrentDocument()).toBe(false);
    expect(
      (document.getElementById("editingVersionDate") as HTMLInputElement).value,
    ).toBe("1700000000000");
  });
});

describe("getRevision", () => {
  it("asks the history of the edited translation for pretty names", async () => {
    const fetchMock = vi.fn(async () =>
      Response.json({ version: "3.1", modified: 1, author: "XWiki.Admin" }),
    );
    vi.stubGlobal("fetch", fetchMock);

    await new XWikiDocument({ language: "fr" }).getRevision("3.1");

    expect(legacyAPI.getRestURL).toHaveBeenCalledWith(
      "translations/fr/history/3.1",
      new URLSearchParams({ prettyNames: "true" }),
    );
  });
});
